jax_unirep/evotuning.py

```diff
@@ -196,6 +194,118 @@ def step(i, state):
+
+    return sum_loss / len(sequences)
+
+
+def fit_manual(
+    params: Dict,
+    train_seqs: List[str],
+    in_val_seqs: List[str],
+    out_val_seqs: List[str],
+    n: int,
+    step_size: float = 0.001,
+    steps_per_print: int = 200,
+) -> Dict:
+    """
+    Return weights fitted to predict the next letter in each sequence.
+
+    The training loop is as follows.
+    Per step in the training loop,
+    we loop over each "length batch" of sequences and tune weights
+    in order of the length of each sequence.
+    For example, if we have sequences of length 302, 305, and 309,
+    over K training epochs,
+    we will perform 3xK updates,
+    one step of updates for each length.
+
+    To get batching of sequences by length done,
+    we call on ``batch_sequences`` from our ``utils.py`` module,
+    which returns a list of sub-lists,
+    in which each sub-list contains the indices
+    in the original list of sequences
+    that are of a particular length.
+
+    :param params: mLSTM1900 parameters.
+    :param train_seqs: The set of training sequences to tune to.
+    :param in_val_seqs: In-domain validation set of sequences,
+        to check for loss on to prevent overfitting.
+    :param out_val_seqs: Out-domain validation set of sequences,
+        to check for loss on to prevent overfitting.
+    :param n: The number of iterations to evotune on.
+    :param step_size: Optimizer learning rate. 
+    :param steps_per_print: Iteration spacing between
+        printing current params and loss. 
+
+    """
+    xs, ys = length_batch_input_outputs(train_seqs)
+
+    init, update, get_params = adam(step_size=step_size)
+    # optimizer_funcs = jit(update), jit(get_params)
+
+    @jit
+    def step(i, state):
+        """
+        Perform one step of evolutionary updating.
+
+        This function is closed inside `fit` because we need access
+        to the variables in its scope,
+        particularly the update and get_params functions.
+
+        By structuring the function this way, we can JIT-compile it,
+        and thus gain a massive speed-up!
+
+        :param i: The current iteration of the training loop.
+        :param state: Current state of parameters from jax.
+        """
+        params = get_params(state)
+        g = grad(evotune_loss)(params, x, y)
+        state = update(i, g, state)
+        return state
+
+    state = init(params)
+    from time import time
+
+    for i in range(n):
+        for x, y in zip(xs, ys):
+            state = step(i, state)
+            params = get_params(state)
+
+        if (i + 1) % steps_per_print == 0:
+
+            # calculate loss for in & out domain validation sets
+            # then print it.
+            # TODO: print to .log file??
+
+            print(
+                "Iteration {0}: in-val-loss={1}, out-val-loss:{2}".format(
+                    i,
+                    avg_loss(in_val_seqs, params),
```

Comment:
For consistency, I think it would be better to use f-string syntax here.

Reply:
@ElArkk & @ivanjayapurna - suggestion below. 

Not sure if it's possible to directly commit from PR page, though worth trying.

```suggestion
            in_val_loss, out_val_loss = avg_loss(in_val_seqs, params)
            print(
                f"Iteration {i}: in-val-loss={in_val_loss}, out-val-loss:{out_val_loss}"
```